BoundedSumKnownN.tex

```diff
@@ -0,0 +1,351 @@
+% Is it correct to say x_i? I mean the i-th element in the multiset
+
+\subsection{Pseudocode in Python}
+We present a simplified Python-like pseudocode of the Rust implementation below. The necessary definitions for the pseudocode can be found at \href{https://www.overleaf.com/project/60d215bf90b337ac02200a99}{``List of definitions used in the pseudocode"}.
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Variable \texttt{n} must be of type \texttt{usize}.
+        \item Type \texttt{T} must have traits \texttt{TotalOrd} and \texttt{Sub(Output=T)}.\footnote{For now, the OpenDP library only implements \texttt{PartialOrd}, but \texttt{TotalOrd} will soon be implemented.}
+\end{itemize}
+\end{itemize}
+
+\subsubsection*{Postconditions}
+\begin{itemize}
+    \item A \texttt{Transformation} is returned (i.e., if a \texttt{Transformation} cannot be returned successfully, then an error should be returned).
+\end{itemize}
+
+\begin{lstlisting}[language=Python, escapechar=|]
+def MakeBoundedSumN(L: T, U: T, n: usize):
+    input_domain = SizedDomain(VectorDomain(IntervalDomain(L, U)), n)
+    output_domain = AllDomain(T)
+    input_metric = SymmetricDistance()
+    output_metric = AbsoluteDistance(T)
+    
+    if L*n < get_min_value(T) or U*n > get_max_value(T): 
+        raise Exception('Invalid parameters') |\label{line:max}|
+    
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*(U-L)/2
+    
+    def function(data: Vec(T)) -> T: |\label{line:fn}|
+        return data.iter().sum() |\label{line:sum}|
+    return Transformation(input_domain, output_domain, function, input_metric, output_metric, stability_relation)
+\end{lstlisting}
+
+\iffalse
+\subsubsection*{Conditions as specified in the pseudocode}
+\begin{itemize}
+    \item \underline{Input domain:} sized domain of all vectors of elements in \texttt{IntervalDomain(L, U)}, where \texttt{L} and \texttt{U} are of type \texttt{T} and stand for \textit{lower bound} and \textit{upper bound}, respectively.\footnote{The compiler infers \texttt{float} from the context, which is why it is omitted.} 
+    
+    \item \underline{Output domain:} domain of all elements of type \texttt{T}.
+    
+    \item \underline{Function:} given $v$ of type \texttt{Vec(T)}, it returns an element of type \texttt{T} which corresponds to the sum of all the vector entries $v_i$ of $v$.
+    
+    \item \underline{Input metric:} symmetric distance.\footnote{As of June 24, SubstituteDistance is no longer a metric part of the OpenDP library.}
+    
+    \item \underline{Output metric:} absolute distance.
+    
+    \item \underline{Stability relation:} for $\din, \dout$ of type \texttt{u32}, the relation returns $\True$ if and only if $\dout \geq \din \cdot (\U-\texttt{L})/2$.
+\end{itemize}
```

Comment:
I don't see this section in the pdf?

Reply:
Is it still necessary to add conditions as specified in pseudo code? even with preconditions and post conditions?

Reply:
It is not in the PDF because I commented it out with the `\iffalse... \fi` LaTeX command (https://tex.stackexchange.com/questions/17816/commenting-out-large-sections) just in case we want to add it back at some point in the future. But no, it is not necessary to add this section now that we have the preconditions, which is why I commented it out.